### DISCUSSION TITLE
Is NIP2 representative of libvips large files implementation and performance?

### DISCUSSION THREAD

**Discussion by Weissrolf** [libvips/libvips#2092] [Q&A]
Hello.

In order to get an impression of libvips performance I tried to load and display large (2+ gb uncompressed, max 32k px) uncompressed and compressed TIFF files via NIP2 (and Photoflow) on a 64-bit W10 PC (9900K).

Both turned out to be very slow for zooming and scrolling, one can literally watch the screen being drawn and sometimes it takes seconds for it to finish.

- Is this due to the image not being loaded completely into memory and all the temporary file-writing happening?

- If so, is this a specific implementation of NIP2 (and Photoflow) or a mandatory design choice/limitation of libvips?

- Is libvips capable of loading these large images completely into memory and thus forgo the constant file-writing and slow zooming/scrolling?

Thanks in advance and nice holidays!

**Comment by jcupitt** [comment]
Hi @Weissrolf,

Yes, by default, libvips keeps files over 100mb on disc. It's supposed to be useful for images which are larger than the amount of RAM you have available -- I often view and manipulate 300,000 by 300,000 pixel images on my laptop. This means that zooming and panning large images in nip2 can need a lot of IO.

The "how libvips opens files" chapter in the docs has some detail on the process:

https://libvips.github.io/libvips/API/current/How-it-opens-files.md.html

You can change the 100mb threshold with the environment variable `VIPS_DISC_THRESHOLD` or with the command-line argument `--vips-disc-threshold`. Set it to eg. `10gb` if you want to keep files under 10gb in memory.

From the API, there's also `vips_image_copy_memory()` to force an image into RAM:

https://libvips.github.io/libvips/API/current/VipsImage.html#vips-image-copy-memory

The win build of nip2 is a bit slower than the linux one due to the heavy use of memory-mapped files. It might be worth checking the speed in linux too, if you're able.

**Comment by Weissrolf** [comment]
Thanks for the quick answer @jcupitt, it's well appreciated.

Is VIPS_DISC_THRESHOLD an OS environmental variable? If so then at least NIPS2 does not seem to care for it. It keeps creating its temp .v file of a size equal to the uncompressed image and does not allocate memory even close to the image size.

That being said, the whole file fits into Windows cache and no reads are performed while I zoom and scroll, but NIPS2 still is very slow at building its view. When I zoom several steps via quick mouse-wheel flick then it takes over a second until NIPS2 even begins building up the new view, which then is build in tiles at visibly slow speed.

Have a look at these screenshot to get an impression of just how slow it is: 

Scrolling: https://i.imgur.com/3yr37Ly.jpg
Zooming: https://i.imgur.com/nv5LwTg.png

Unfortunately I cannot tell if this is a limitation of NIP2 or of libvips. This is why I looked at Photoflow for comparison, but it showed the same problems. Are there any other applications making use of libvips that demonstrate its performance potential for loading/decoding/decompression and display of large images?

Since I am starting from scratch I would have liked to see libvips in action before putting too much time and effort into it.

**Comment by jcupitt** [comment]
Ah true, I've not tried disabling the disc cache for nip2. Now I think about it, I recall it has its own system.

Image display is asynchronous and threaded, so the delay on rapid zoom is the threadpool and pipeline being built, cancelled and rebuilt several times. Try performing an expensive operation on a large image (Filter / Morphology / Custom morphology with the iterations turned up?) and then zooming and panning on the result -- it might make the process of computation clearer. You could also try changing the image display controls -- View / Toolbar / Display control in an image window, then dragging one of the sliders.

Programs like photoshop and gimp maintain pyramids to keep panning of large, zoomed-out images fast, but nip2 does not. No pyramids means operations are quick (try performing several Canny filters in a row, then adjusting a parameter on one of the earlier ones), but zooming out is slow.

Did you see vipsdisp?

https://github.com/libvips/vipsdisp-tiny

It's a minimal image viewer in ~300 lines of C. It's supposed to be useful for looking at the libvips async repaint model and how the redraw system works. It might help evaluation.

There's a more complete image viewer here:

https://github.com/jcupitt/vipsdisp

It works, though it's not quite finished. It's supposed to eventually be the image display window for nip3 (when I get around to it).

**Comment by jcupitt** [comment]
I meant to say: if fast pan on large, zoomed-out images is important for your application, you'll probably need a pyramid system. You could still use libvips, of course, but you'd have to implement the pyramid code yourself. A simple pyramid structure is very little code (you can use libvips for the shrinking, obviously), but it would make operations slower.

**Comment by Weissrolf** [comment]
Thanks again! At this point I am mainly interested in fast loading and smooth viewing while keeping a responsive UI. OpenCV (+QT) based Nomacs can load and display my somewhat non-standard uncompressed 2.9 gb TIFF file in 2 seconds, despite using only a single thread. (Non-standard, because it is only 2.17 gb of data in a larger file. No idea how I created that, but likely using Gigapixel or Acdsee.)

Compressed TIFF and PNG images seem to be a problem for all image viewers/editors out there, because they (de)compress using only a single thread. At least for ZIP/Deflate I find this rather surprising, but I admittedly lack any knowledge about the internals of image compression.

One of the reasons I am here is that I grow tired of twiddling thumbs sitting in front of expensive hardware while most of its resources sit idle and UIs being unresponsive (animated circles, balls, "not responding").

Again, happy holidays! My kids are building Lego and puzzles right now. ;)

**Comment by jcupitt** [comment]
vipsdisp will load and display a big uncompressed strip tiff at about that speed, and a 30,000 x 30,000 jpg is similar. Plus it only needs ~150mb of memory to do it, since the image is kept on disc.

Any size tiled TIFF (with or without compression) opens instantly, since it doesn't need to decompress to get random access to pixels. Zoom out will be slow though. PNG is just horribly slow all the time for all viewers :( 

Yes, happy Christmas!

**Comment by Weissrolf** [comment]
Does vipsdisp allow to be compiled to a Windows binary and does its basic display output differ from NIP2?

I noticed that you specifically mentioned panning zoomed "out" images. But with NIP2 shows panning artifacts even at 100%?!

https://i.imgur.com/rJjDbCL.jpg

Using NIP2 I saved and then opened a tiled compressed TIFF version of my test-file. It opens very fast in NIP2, but zooming out is really slow then, with tiles slowly building bottom to top. None of my 16 logical CPU cores is even close to being maxed out and IO throughput is only 125-150 mb/s despite the data residing in cache (and my M2 being able to read at 2-3 gb/s).

The same basically happens with the uncompressed TIFF, but comparably faster. So I wonder where the bottleneck is when neither CPU nor IO seem maxed out?

This is not a critique, I am merely trying to clarify if this is expected performance and by design?!

**Comment by jcupitt** [comment]
I've not tried building vipsdisp on Windows, but it shouldn't be too hard, I'd think -- you'd need a gtk3 binary to go with the vips binary, but nothing else.

nip2 on Windows does artefact a bit. I mostly finished nip2 in about 2005 I think, and it uses gtk2 and vips7. Since then of course gtk3 and vips8 have come along. vips8 is fairly compatible with vips7, but gtk3 has a radically different (and much better) gpu-friendly drawing model.

I've been looking at updating nip2 to vips8 / gtk3 for a while, and the new drawing system is the tricky part. vipsdisp is supposed to be that component: all the drawing is redone for the new framework. It ought to be a lot better.

Tiled compressed TIFF is an interesting case. libtiff handles JPEG decompression for you, but it's all single-threaded, unfortunately (libtiff was designed in the days before threads). nip2 as a design choice has almost no data structures supporting zooming, other than a tile cache, so while latency is very low, a lot of computation must happen during display update, especially when zoomed out.

openslide is a good point of comparison. These images are often in a tiled TIFF container, but with jp2k-compressed tiles. openslide puts tile decode outside the read lock, so you can decompress tiles in parallel. If you try downloading one of their demo images:

http://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/JP2K-33003-2.svs

You should see all your cores light up as you pan.

You could also try doing some processing on your image. Try running several canny filters on it (for example), then panning the final image. You should see load on all cores.

**Comment by jcupitt** [comment]
I made a very quick screen recording of vipsdisp (easier than making a win build) showing performance with a large jpg, tiled compressed tiff, and openslide:

http://www.rollthepotato.net/~john/simplescreenrecorder-2020-12-25_09.35.53.mp4

**Comment by jcupitt** [comment]
Oh heh that recording flickers a bit now I look at it, but that's just from the screen recorder I used.

**Comment by Weissrolf** [comment]
Hey John, thanks for taking time to explain things and even do a video. The video demonstrates various things I listed for NIP2 earlier, like CPU load not maxing out with the JPEG and TIFF while we still sit and watch the image being build up (2:07 min).

My current understanding of the situation is:

- Ongoing processing of image data during pans and zooms is a design goal of libvips in order to keep memory footprint small and allow to display images whose size may (or may not) exceed available memory. For this goal a copy of the (decoded?) image data is written to the harddrive, but no off-screen data is kept in memory.
- Because no off-screen data is kept in memory, image data has to be recalculated every time it is revisited after being shifted off-screen.
- On-screen image data is (decoded again? and) drawn in asynchronous tiles.
- Panning the on-screen display to reveal off-screen data leads to already present on-screen data being drawn in repeated patterns (or "smeared") across the screen until the new areas are finished being decoded and drawn (=no blank border appearing).
- Thread/pipeline initialization overhead seems to bottleneck the asynchronous decoding process. As a result we see (and wait for) on-screen image data being build up, even though neither CPU nor IO are maxed out.

Questions:

- Is it possible to tell libvips to optionally keep already decoded image data in memory instead of reprocessing it once it went off-screen? And is it possible to work entirely in memory for images small enough to fit instead of creating temporary files? Would this speed up panning of areas already decoded to memory?
- Is it possible to limit the number of threads (overhead) being used for these relatively "small" images where we know that a complete 2 gb uncompressed TIFF can be decoded in 2 seconds by a single thread even? By limit I mean anything in between 1 and maximum #cores threads. Would this speed up zooming in those cases were we don't see CPU/IO being maxed out?

**Comment by jcupitt** [comment]
Hi again, on your points:

- Processing during pan and zoom is a design goal of *nip2* (and vipsdisp) rather than libvips. Fast pan and zoom needs an image pyramid, and nip2 (as you say) chooses not maintain one of these, since they need a lot of resources to build. I have nip2 workspaces with over 10,000 images (doing things like pulmonary PET CT modelling) and maintaining a pyramid for every one would not be possible.
- The nip2 tile cache is sized large enough to hold two desktops screens (as I recall). You can slide tiles off the edge of the window and back again for some distance. You can adjust the size: https://libvips.github.io/libvips/API/current/libvips-conversion.html#vips-tilecache
- I experimented with blanking out off-screen tiles, but decided I actually preferred the smearing. It makes it more obvious when tiles arrives, and what bits of the screen have not been drawn yet.
- vipsdisp handles pipeline start and cancel much better. The long pauses with the openslide image are vipsdisp waiting for tiles which it asked for but no longer needs to calculate. You'll see the CPU is maxed out in those periods.

Your questions:

- Yes, it's flexible, you can set any size tile cache, and you can decode to memory rather than disc.
- vipsdisp is doing this now -- it opens that 2gb JPG image in about 2s, and zoom and pan is quick. Uncompressed TIFF is about the same speed.

**Comment by Weissrolf** [comment]
Great clarifications, it helps get a clearer picture of what is library design vs. what is application design. The hint about the tile-cache being 2 desktops large also helped doing better tests to see processing vs. drawing quirks.

I noticed that the tile-cache is relevant per current zoom level. If I understand correctly then it should be possible to keep the 100% image data completely in memory and then use the tile-cache for panning non 100% zoom levels?!

Concerning decompression: Why it not possible to compress/decompress Deflate via multiple threads? Is this a limitation of Deflate?

7zip only uses a single thread when Zip/Deflate is used, but Winrar seems to use multi-threading even for Zip/Deflate, albeit without being able to max out my CPU cores.

**Comment by jcupitt** [comment]
Threaded decompress depends on restart markers. 7z writes a restart marker every ~10kb (I think), so each chunk can be decompressed independently. PNG (for example) supports restart markers in theory, but no one ever writes them (ie. the whole of the image data is a single chunk), so you can't usually thread PNG read.

You can write tiled TIFF with deflate compression, which would then allow threaded decompress, but libtiff calls libz inside its read lock, so it ends up not being useful :( libtiff has support for reading the compressed tile data, so you could in theory do the decompress in parallel yourself, though I think very few developers take the trouble.

Yes, nip2 has one tile cache for each image view window and drops it on each change of zoom level. It would be a simple change for it to keep perhaps the most recent 10 tile caches (or the 10 with the largest zoom out?)  -- that might be an easy way to speed up zooming.

**Comment by Weissrolf** [comment]
If PNG already supports this then it is even more offensive that it is not widely implemented. We are still living in the digital stone ages when it comes to such things.

Uncompressed TIFF is not always the answer, because of the 4 gb file size limit and slower network access to such large files. 

TIFF Deflate can be problematic, not only because of different compression ratios used by different software (not user controlled), but because some software cannot read the files created by other software. It took months until Adobe fixed Photoshop not being able to read compressed TIFF files coming from Topaz Gigapixel or ACDsee.

PNG solves the latter problem, but its single-threaded and seemingly mandatory compression implementation is just slow and thus annoying. 

Saving my testfile from Photoshop as uncompressed TIFF takes 3 seconds, this turns to around 3 minutes for Deflate compressed TIFF, but 5.5 minutes for PNG (largest file/fastest compression). How is that ok for modern multi-core computing?

**Comment by Weissrolf** [comment]
Hello John

It seems that there are some efforts to make both TIFF (libtiff) and PNG Deflate compression faster/multi-threaded.

TIFF: https://gitlab.com/libtiff/libtiff/-/merge_requests/158
PNG: https://github.com/brion/mtpng

There is also this commercial TIFF output plugin for the Harlequin RIP that claims to be multi-threaded: http://www.hamillroad.com/products/lightning-tiff/

**Comment by jcupitt** [comment]
There's libspng too -- it's a reimplementation of libpng with a much better API and better performance. If you build with zlib-ng as well, you can get a nice speedup:

https://github.com/zlib-ng/zlib-ng

https://github.com/randy408/libspng

openslide does multithreaded TIFF decode too, of course.

**Comment by jcupitt** [comment]
I didn't know about mtpng, thanks for the link!

**Comment by Weissrolf** [comment]
According to the following benchmark zlib and zlib-ng seem to mostly perform the same, though, while libdeflate seems to offer real performance gains.

https://quixdb.github.io/squash-benchmark/unstable/#ratio-vs-compression

This benchmark gives zlib-ng some more love: https://sites.google.com/site/powturbo/home/web-compression

One letdown about mtpng is that its possible performance gains for decompression seem limited due to the streaming vs. block oriented nature.

Libdeflate does not offer good streaming yet either, but I don't know what this means for TIFF compression/decompression. 

Are there projects (maybe even with binary demos) that implement libspng? Any plans for vips in that regard?

**Comment by jcupitt** [comment]
Yes, libvips 8.10 defaults to libspng for decode. libspng encode is still in development, but we hope to switch over when it's done.

I think the next big thing is supposed to be AVIF. It's a much more modern codec, about 2x better than jpg, has a good lossless mode, and no patent hazards. It's been designed with GPU acceleration in mind as well. Perhaps it'll fix all our problems.

**Comment by Weissrolf** [comment]
Good to know. Spgn is not listed in the optional dependencies, so I did not know it was already half-way there. :)

And yes, it's time for a more modern codec to take over. This also includes images within PDF files, which are often a bottleneck when viewing image-heavy documents.

**Comment by jcupitt** [comment]
Oooop, you're right, I forgot to add a note about libspng to the README. I've put it in now, thanks.

**Comment by Weissrolf** [comment]
I learned a lot, thanks for that.

**Comment by randy408** [comment]
> If PNG already supports this then it is even more offensive that it is not widely implemented. We are still living in the digital stone ages when it comes to such things.

I have a basic draft of a PNG extension to support this, in theory it should be backward-compatible with minimal overhead, but obviously encode support has to come first.

**Comment by jcupitt** [comment]
A quick update since things have changed a bit since we ha this interesting discussion:

1. libvips now supports threaded TIFF tile decode for JPEG and JP2K compression. It'd be nice to add threaded deflate decompress too.
2. TIFF write is still single-threaded, sadly. It wouldn't be too much work to fix this, but no one's found the time yet.
3. vipsdisp is now a gtk4 program and has a super fancy new drawing system. It keeps a pyramid of recent tiles in memory, and composites them to the screen using your GPU. Press "d" to toggle a debug mode that lets you watch tiles being swapped in and out. Because window update and tile generation are decoupled, you can pan and zoom at 60fps, with tiles swapping in and out in the background as libvips generates them. There's a binary on flathub so it should be easy to try out. https://flathub.org/apps/details/org.libvips.vipsdisp
